Rubberduck.Parsing/VBA/COMReferenceManagerBase.cs

```diff
@@ -0,0 +1,269 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using NLog;
+using System.Collections.Concurrent;
+using Rubberduck.VBEditor;
+using System.Runtime.InteropServices;
+using System.IO;
+using Rubberduck.Parsing.ComReflection;
+
+namespace Rubberduck.Parsing.VBA
+{
+    public abstract class COMReferenceManagerBase : ICOMReferenceManager
+    {
+        private const string rubberduckGUID = "{E07C841C-14B4-4890-83E9-8C80B06DD59D}";
+
+        protected readonly RubberduckParserState _state;
+        protected readonly IParserStateManager _parserStateManager;
+        private readonly string _serializedDeclarationsPath;
+
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+
+        public COMReferenceManagerBase(RubberduckParserState state, IParserStateManager parserStateManager, string serializedDeclarationsPath = null)
+        {
+            if (state == null) throw new ArgumentNullException(nameof(state));
+            if (parserStateManager == null) throw new ArgumentNullException(nameof(parserStateManager));
+
+            _state = state;
+            _parserStateManager = parserStateManager;
+            _serializedDeclarationsPath = serializedDeclarationsPath
+                ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Rubberduck", "declarations");
+        }
+
+
+        private bool _lastRunLoadedReferences;
+        public bool LastRunLoadedReferences
+        {
+            get
+            {
+                return _lastRunLoadedReferences;
+            }
+        }
+
+        private bool _lastRunUnloadedReferences;
+        public bool LastRunUnloadedReferences
+        {
+            get
+            {
+                return _lastRunUnloadedReferences;
+            }
+        }
+
+        private readonly HashSet<ReferencePriorityMap> _projectReferences = new HashSet<ReferencePriorityMap>();
+        public ICollection<ReferencePriorityMap> ProjectReferences
+        {
+            get
+            {
+                return _projectReferences.ToHashSet();
+            }
+        }
+
+
+        protected abstract void LoadReferences(List<IReference> referencesToLoad, ConcurrentBag<IReference> unmapped, CancellationToken token);
+
+
+        public void SyncComReferences(IReadOnlyList<IVBProject> projects, CancellationToken token)
+        {
+            _lastRunLoadedReferences = false;
+            _lastRunUnloadedReferences = false;
+
+            var unmapped = new ConcurrentBag<IReference>();
+
+            var referencesToLoad = GetReferencesToLoadAndSaveReferencePriority(projects);
+
+            _state.OnStatusMessageUpdate(ParserState.LoadingReference.ToString()); //Why do we use this method to change the displayed state?
```

Comment:
`OnStatusMessageUpdate` is a hack - because RD commandbar status label used to be solely commanded from the ParserState (gosh, writing this down gives me chills), so `OnStatusMessageUpdate` was devised to update the status label without changing any parser state.